group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformHeterogeneousAssignmentBuilder.java

```diff
@@ -157,716 +240,779 @@ public GroupAssignment build() {
-                    ')';
+            // Find the least loaded subscriber.
+            int leastLoadedMemberIndex = memberAssignmentBalancer.nextLeastLoadedMember();
+
+            // If the most and least loaded subscribers came from member ranges with partition
+            // counts that differ by one or less, then the members are balanced and we are done.
+            if (memberAssignmentBalancer.isBalanced()) {
+                // The topic is balanced.
+                break;
             }
-        }
 
-        /**
-         * Initializes an AssignmentManager, setting up the necessary data structures.
-         */
-        public AssignmentManager(
-            SubscribedTopicDescriber subscribedTopicDescriber
-        ) {
-            groupSpec.memberIds().forEach(memberId -> {
-                int maxSize = groupSpec.memberSubscription(memberId).subscribedTopicIds().stream()
-                    .mapToInt(subscribedTopicDescriber::numPartitions)
-                    .sum();
-
-                MemberAssignmentData memberAssignmentData = membersWithAssignmentSizes
-                    .computeIfAbsent(memberId, MemberAssignmentData::new);
-                memberAssignmentData.maxAssignmentSize = maxSize;
-                memberAssignmentData.currentAssignmentSize = 0;
-            });
+            // Reassign the partition.
+            assignPartition(topicId, partition, leastLoadedMemberIndex);
+            reassignedPartitionCount++;
         }
 
-        /**
-         * @param memberId       The member Id.
-         * @return The current assignment size for the given member.
-         */
-        private int targetAssignmentSize(String memberId) {
-            MemberAssignmentData memberData = this.membersWithAssignmentSizes.get(memberId);
-            if (memberData == null) {
-                LOG.warn("Member Id {} not found", memberId);
-                return 0;
-            }
-            return memberData.currentAssignmentSize;
-        }
+        return reassignedPartitionCount;
+    }
 
-        /**
-         * @param memberId      The member Id.
-         * @return The maximum assignment size for the given member.
-         */
-        private int maxAssignmentSize(String memberId) {
-            MemberAssignmentData memberData = this.membersWithAssignmentSizes.get(memberId);
-            if (memberData == null) {
-                LOG.warn("Member Id {} not found", memberId);
-                return 0;
+    /**
+     * Restores topic partitions to their previous owners if it improves balance.
+     *
+     * @param topicId                  The topic to rebalance.
+     * @param numPartitions            The number of partitions in the topic.
+     * @param previousPartitionOwners  The previous owners of the partitions.
+     * @param partitionOwners          The current owners of the partitions.
+     * @return the number of partitions reassigned.
+     */
+    private int balanceTopicRestoringStickiness(
```

Comment:
I wonder if this step really make a big different. Does it based on your experience?

Reply:
This step is a rough translation of something the previous implementation did. The only time it does anything is if a previous balancing iteration moved partitions away from a member and now wants to move them back. For this to happen, subscriptions must be overlapping and non-equal.

eg. Consider a case with 2 topics and 3 consumers, where topic 1 is subscribed to by consumer 1 and 2 and topic 2 is subscribed to by consumer 2 and 3. Let's say that consumer 1 has 10 partitions, 2 has 15 and 3 has 5. We might balance topic 1 first and move partitions from consumer 2 -> 1, so that they both have 12-13 partitions. Then we might balance topic 2 and move partitions from consumer 2 -> 3, so that they both have 9 partitions. The next iteration on topic 1 would move partitions back from consumer 1 -> 2.

Since subscription patterns like these are uncommon, I'm actually tempted to remove this step.

Reply:
Yeah, I agree. This was my first though too.

Reply:
Do you have an idea of the gain if we remove it?

Reply:
I took a look at profiles for the smallest and largest full and incremental cases and it's pretty small, consistently less than 0.50%.

Reply:
Ack. I am fine with removing it if it is not really useful in the end.

Reply:
I've removed it and `previousAssignmentPartitionOwners` in the latest version.